src/encoding/byte_stream_split/decoder.rs

```diff
@@ -0,0 +1,61 @@
+use std::marker::PhantomData;
+use crate::error::Error;
+use crate::types::NativeType;
+
+/// Decodes according to [Byte Stream Split](https://github.com/apache/parquet-format/blob/master/Encodings.md#byte-stream-split-byte_stream_split--9).
+/// # Implementation
+/// This struct does not allocate on the heap.
+#[derive(Debug)]
+pub struct Decoder<'a, T: NativeType> {
+    values: &'a [u8],
+    buffer: Vec<u8>,
+    num_elements: usize,
+    current: usize,
+    element_size: usize,
+    element_type: PhantomData<T>
+}
+
+impl<'a, T: NativeType> Decoder<'a, T> {
```

Comment:
Can we maybe make this `const` on `mem::size_of::<T>`? Then we can use an `array: [u8; Size]` instead of a heap allocation.

Reply:
Sorry, I'm new to rust, but if I understand you correctly, I'd need to use:
```
#![feature(generic_const_exprs)]
```
and add a where clause, akin to:
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=d8088113ef4552d7f96b987ec4d31c95

Are we ok to depend on non-stable features, or is there a better way to do this?

Reply:
If using an unstable feature isn't okay, an alternative approach could be adding an extra type parameter, but this does make the consumer API a little more verbose. Eg: https://github.com/adamreeve/parquet2/commit/a5138455b29c0de7634a9c66437cfbf9a1767138

Reply:
I think we need a new `const` type parameter for that. I can take a look later if you want.

Reply:
Yes please, that would be great.